gtsfm/retriever/netvlad_retriever.py

```diff
@@ -0,0 +1,261 @@
+"""Retriever implementation which provides a NetVLAD global image descriptor to suggest potential image pairs.
+
+Note: Similarity computation based off of Paul-Edouard Sarlin's HLOC:
+Reference: https://github.com/cvg/Hierarchical-Localization/blob/master/hloc/pairs_from_retrieval.py
+https://openaccess.thecvf.com/content_cvpr_2016/papers/Arandjelovic_NetVLAD_CNN_Architecture_CVPR_2016_paper.pdf
+
+Authors: John Lambert
+"""
+
+import math
+import os
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List, Optional, Tuple
+
+import dask
+import matplotlib.pyplot as plt
+import numpy as np
+import torch
+from dask.delayed import Delayed
+
+import gtsfm.utils.logger as logger_utils
+from gtsfm.loader.loader_base import LoaderBase
+from gtsfm.retriever.retriever_base import RetrieverBase
+from gtsfm.frontend.cacher.global_descriptor_cacher import GlobalDescriptorCacher
+from gtsfm.frontend.global_descriptor.netvlad_global_descriptor import NetVLADGlobalDescriptor
+
+
+logger = logger_utils.get_logger()
+
+PLOT_SAVE_DIR = Path(__file__).parent.parent.parent / "plots"
+
+MAX_NUM_IMAGES = 10000
+
+
+@dataclass
+class SubBlockSimilarityResult:
+    i_start: int
+    i_end: int
+    j_start: int
+    j_end: int
+    subblock: torch.Tensor
+
+
+class NetVLADRetriever(RetrieverBase):
+    def __init__(self, num_matched: int, blocksize: int = 10) -> None:
+        """
+        Args:
+            num_matched: number of K potential matches to provide per query. These are the top "K" matches per query.
+            blocksize: size of matching sub-blocks when creating similarity matrix.
+        """
+        self._num_matched = num_matched
+        self._global_descriptor_model = GlobalDescriptorCacher(global_descriptor_obj=NetVLADGlobalDescriptor())
+        self._blocksize = blocksize
+        self._min_score = 0.1
+
+    def create_computation_graph(self, loader: LoaderBase) -> Delayed:
+        """Compute potential image pairs.
+
+        Args:
+            loader: image loader. The length of this loader will provide the total number of images
+                for exhaustive global descriptor matching.
+
+        Return:
+            Delayed task that evaluates to a list of (i1,i2) image pairs.
+        """
+        return self.run(loader=loader)
+
+    def run(self, loader: LoaderBase, visualize: bool = True) -> Delayed:
+        """Compute potential image pairs.
+
+        Args:
+            loader: image loader. The length of this loader will provide the total number of images
+                for exhaustive global descriptor matching.
+            visualize:
+
+        Return:
+            Delayed task which evaluates to a list of (i1,i2) image pairs.
+        """
+        num_images = len(loader)
+        sim = self.compute_similarity_matrix(loader, num_images)
+        return dask.delayed(self.compute_pairs_from_similarity_matrix)(sim=sim, loader=loader, visualize=visualize)
```

Comment:
not sure how this is done in other places (@ayushbaid )
isnt the Dask.delayed supposed to be created in create_computation_graph? so that run() returns the actual type.

Reply:
Lets have the run method not use any dask delayed. We can duplicate the logic in create_computation_graph if we like. The reason is that non-dask pathways can be very useful for large datasets (e.g. Hilti).